### ISSUE TITLE
Stanford Online Product training scheme

### ISSUE THREAD

**Issue by elias-ramzi** [KevinMusgrave/powerful-benchmarker#81]
Hi Kevin, thanks a lot for this repo :)

I'm working on Standford Online Product and I would like to train networks with the same scheme as deep metric learning paper. I am not sure how to proceed for the two following issues :

1. Using the train/test split already created, and have class disjoint train/val sets
2. Using the FastAP sampling scheme : you first sample two categories, and then you sample with MPerClassSampler within those two categories (with batch size // 2)

I am new to the repository but I would be glad to help you on those issues

Thank you !

**Comment by KevinMusgrave** [comment]
> 1. Using the train/test split already created, and have class disjoint train/val sets

Could you give an example? I'm not 100% sure what you mean.

Currently, this repo will split the dataset into 50% trainval and 50% test. Then it will split the trainval into 75% train and 25% val. The trainval is class disjoint from the test set, and the train and val sets are also class disjoint.

> 2. Using the FastAP sampling scheme : you first sample two categories, and then you sample with MPerClassSampler within those two categories (with batch size // 2)

This should probably go into pytorch-metric-learning. Here's the relevant discussion: https://github.com/KevinMusgrave/pytorch-metric-learning/issues/105

**Comment by elias-ramzi** [comment]
The train and test sets are already defined for SOP. They were created by sampling 50% of classes for each categories. I think with the way the SOP dataset is implemented in you repo the train/test splits are almost correct. But I am not sure on how to use the exact pre-defined train/test splits.
It would also be interesting to have a split manager that takes into account the categories.


Ok I think I will do a PR with the implementation based on : https://github.com/kunhe/Deep-Metric-Learning-Baselines/blob/master/datasets.py for the HierarchicalSampler.

**Comment by KevinMusgrave** [comment]
Re: SOP, I'm not sure if there will be an easy way to do that with this library. To be honest, the code is a bit of a mess right now. I started a major rewrite last summer but didn't have time to get it to a clean state.

**Comment by KevinMusgrave** [comment]
I'll take a look at the code later this week to see if there's a straightforward to implement your idea.

**Comment by elias-ramzi** [comment]
Ok thank you !

I opened a PR on pytorch_metric_learning for the sampler : https://github.com/KevinMusgrave/pytorch-metric-learning/pull/303
